File: python/features/steps/issue_close_delete_steps.py
"""Behave steps for issue close and delete."""

from __future__ import annotations

from behave import then

from features.steps.shared import load_project_directory
from kanbus.project import find_project_local_directory


def _issue_path_for_identifier(context: object, identifier: str):
    project_dir = load_project_directory(context)
    path = project_dir / "issues" / f"{identifier}.json"
    if path.exists():
        return path
    local_dir = find_project_local_directory(project_dir)
    if local_dir is not None:
        local_path = local_dir / "issues" / f"{identifier}.json"
        if local_path.exists():
            return local_path
    return project_dir / "issues" / f"{identifier}.json"


@then('issue "{identifier}" should not exist')
def then_issue_not_exists(context: object, identifier: str) -> None:
    path = _issue_path_for_identifier(context, identifier)
    assert not path.exists(), f"Expected issue {identifier} to be deleted"


@then('issue "{identifier}" should exist')
def then_issue_exists(context: object, identifier: str) -> None:
    path = _issue_path_for_identifier(context, identifier)
    assert path.exists(), f"Expected issue {identifier} to exist"

File: rust/features/steps/issue_close_delete_steps.rs
use std::path::PathBuf;

use cucumber::then;

use crate::step_definitions::initialization_steps::KanbusWorld;
use kanbus::file_io::{find_project_local_directory, load_project_directory};

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

fn issue_path_for_identifier(world: &KanbusWorld, identifier: &str) -> PathBuf {
    let project_dir = load_project_dir(world);
    let path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    if path.exists() {
        return path;
    }
    if let Some(local_dir) = find_project_local_directory(&project_dir) {
        let local_path = local_dir.join("issues").join(format!("{identifier}.json"));
        if local_path.exists() {
            return local_path;
        }
    }
    project_dir
        .join("issues")
        .join(format!("{identifier}.json"))
}

#[then(expr = "issue {string} should not exist")]
fn then_issue_not_exists(world: &mut KanbusWorld, identifier: String) {
    let path = issue_path_for_identifier(world, &identifier);
    assert!(
        !path.exists(),
        "Expected issue {} to be deleted",
        identifier
    );
}

#[then(expr = "issue {string} should exist")]
fn then_issue_exists(world: &mut KanbusWorld, identifier: String) {
    let path = issue_path_for_identifier(world, &identifier);
    assert!(path.exists(), "Expected issue {} to exist", identifier);
}
